edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/Handlers/TwinHandler.cs

```diff
@@ -0,0 +1,350 @@
+                                                    twin.Body);
+                }
+                catch (Exception e)
+                {
+                    Events.FailedToSendTwinUpdateMessage(e);
+                    result = false;
+                }
+
+                if (result)
+                {
+                    Events.TwinUpdate(identity.Id, statusCode, correlationId, twin.Body.Length);
+                }
+                else
+                {
+                    Events.TwinUpdateFailed(identity.Id, statusCode, correlationId, twin.Body.Length);
+                }
+            }
+            else
+            {
+                Events.TwinUpdateIncompete(identity.Id);
+            }
+        }
+
+        public async Task SendDesiredPropertiesUpdate(IMessage desiredProperties, IIdentity identity)
+        {
+            if (!desiredProperties.SystemProperties.TryGetValue(SystemProperties.Version, out var version))
+            {
+                Events.DesiredPropertiesUpdateIncompete(identity.Id);
+                return;
+            }
+
+            bool result;
+            try
+            {
+                result = await this.connector.SendAsync(
+                                                GetDesiredPropertiesUpdateTopic(identity, version),
+                                                desiredProperties.Body);
+            }
+            catch (Exception e)
+            {
+                Events.FailedToSendDesiredPropertiesUpdateMessage(e);
+                result = false;
+            }
+
+            if (result)
+            {
+                Events.DesiredPropertiesUpdate(identity.Id, version, desiredProperties.Body.Length);
+            }
+            else
+            {
+                Events.DesiredPropertiesUpdateFailed(identity.Id, version, desiredProperties.Body.Length);
+            }
+        }
+
+        Task<bool> HandleTwinGet(Match match, MqttPublishInfo publishInfo)
+        {
+            return this.HandleUpstreamRequest(
+                        async (proxy, rid) =>
+                        {
+                            await proxy.SendGetTwinRequest(rid);
+                        },
+                        match,
+                        publishInfo);
+        }
+
+        Task<bool> HandleUpdateReported(Match match, MqttPublishInfo publishInfo)
+        {
+            return this.HandleUpstreamRequest(
+                        async (proxy, rid) =>
+                        {
+                            var message = new EdgeMessage.Builder(publishInfo.Payload).Build();
+                            await proxy.UpdateReportedPropertiesAsync(message, rid);
+                        },
+                        match,
+                        publishInfo);
+        }
+
+        async Task<bool> HandleUpstreamRequest(Func<IDeviceListener, string, Task> action, Match match, MqttPublishInfo publishInfo)
+        {
+            var id1 = match.Groups["id1"];
```

Comment:
This seems very loosely coupled. I think a better pattern is - when you do the match, you get the parameters into an object, and pass that object instead.

Reply:
removed the helper function, so these are not passed as parameters anymore